### PR TITLE
pdnsd service: init

### PR DESCRIPTION
**Pull request by ghost**
I added an '--enable-ipv6' config flag over the default package, and tested with the unstable channel at 77f8f35d.

### CODE REVIEW

File: nixos/modules/services/networking/pdnsd.nix [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+          user = mkOption {
+            type = types.str;
+            default = "pdnsd";
+            description = "User account under which pdnsd runs.";
+          };
+
+          cacheDir = mkOption {
+            type = types.str;
+            default = "/var/cache/pdnsd";
+            description = "Directory holding the pdnsd cache";
```
Comment (jagajaga) [inline nixos/modules/services/networking/pdnsd.nix]: Please add dot at the end.

Comment (nfjinjing) [inline nixos/modules/services/networking/pdnsd.nix]: Thanks, dot added!

File: nixos/modules/services/networking/pdnsd.nix [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    };
+
+    systemd.services.pdnsd =
+      { wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        preStart =
+          ''
+            mkdir -p ${cfg.cacheDir}
+            touch ${cfg.cacheDir}/pdnsd.cache
+            chown -R pdnsd ${cfg.cacheDir}
```
Comment (jagajaga) [inline nixos/modules/services/networking/pdnsd.nix]: Maybe you should chown it as services.pdbsd.user?

Comment (jagajaga) [inline nixos/modules/services/networking/pdnsd.nix]: Or why do you need this config option (user)?

Comment (nfjinjing) [inline nixos/modules/services/networking/pdnsd.nix]: I don't need it. I was copying the nginx service file, and in it user seems to be configurable, so maybe I should leave it just as pdnsd?

Comment (nfjinjing) [inline nixos/modules/services/networking/pdnsd.nix]: OK, I removed the user config option.

File: nixos/modules/services/networking/pdnsd.nix [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+with lib;
+
+let
+  cfg = config.services.pdnsd;
+  pdnsd = pkgs.lib.overrideDerivation pkgs.pdnsd
+           (attrs:
+             {
+               configureFlags =
+                 [ "--enable-ipv6"
```
Comment (fpletz) [inline nixos/modules/services/networking/pdnsd.nix]: We should enable IPv6 by default in the `pdnsd` package. Supporting non-legacy IP versions is a must.

Comment (nfjinjing) [inline nixos/modules/services/networking/pdnsd.nix]: I think that would be a cleaner solution. I added another two commits to this PR reflecting this update on the default `pdnsd` package.

**Comment by abbradar** [comment]
LGTM, though I haven't tested this by myself. Thanks! One minor issue left is that cache directory with spaces in it would break because of no quoting in `preStart` script. I'll merge this in a few days (to leave a window for others to comment) -- if noone beats me to it before.

**Comment by nfjinjing** [comment]
Speaking of testing, what I had to do to produce a desired `resolve.conf` was:

1. ignoring DNS from DHCP
2. manually specifying a nameserver in network config

example configuraion:

    networking =
        { 
          dhcpcd.extraConfig =
            ''
              nooption domain_name_servers
              nohook resolv.conf
            '';
            nameservers = [ "::1" ];
        };

**Comment by abbradar** [comment]
Hm, perhaps we want this in the module?

**Comment by nfjinjing** [comment]
According to #12550, users might want to choose whether to use DNS from a DHCP. I'm not sure how to encapsulate this option.

**Comment by abbradar** [comment]
If I understand correctly, `networking.nameservers = [ "::1" ]` would be desired in all cases (to actually use pdnsd). Let's leave the other option out for now, while #12550 is discussed.

**Comment by abbradar** [comment]
Oh, sorry, I got distracted and haven't understood that the issue is closed and the solution is found already ~_^. We may still want this in manual, though -- but this is out of scope of this PR.

What do you think of adding `network.nameservers` to the module?

**Comment by nfjinjing** [comment]
If we expose the `server_ip` option of the global directory of `pdnsd.conf`, then we can fill it in `network.nameservers`. That sounds like a sensible thing to do.

**Comment by nfjinjing** [comment]
Oh about if `::1` should always be the case,  `pdnsd` can also be used as an upstream server for `dnsmasq`, I think, so it might not necessarily listen on `53` on the localhost.

**Comment by nfjinjing** [comment]
Sorry I was confused. What a user put into `network.nameservers` really depends on one's network stack and usage. It could be `127.0.0.1` for IPv4 or `::1` for IPv6; "0.0.0.0" and "::" to share the service. So the value probably should not be fixed.

**Comment by abbradar** [comment]
Well, indeed, options are many so let's leave it as-is. We can always capture most common configurations with additional parameters (`service.pdnsd.useLocally`) later.

**Comment by nfjinjing** [comment]
Sorry about previous comment, `resolve.conf` and `server_ip` are the same only when used locally, so we can't really abstract that out.

**Comment by abbradar** [comment]
Notice that I've changed service UID/GID -- those were already taken.

**Pull request merged**